vertx-mysql-client/src/main/java/io/vertx/mysqlclient/impl/MySQLRowDesc.java

```diff
@@ -18,31 +18,93 @@
 
 import io.vertx.mysqlclient.impl.datatype.DataFormat;
 import io.vertx.mysqlclient.impl.protocol.ColumnDefinition;
+import io.vertx.sqlclient.desc.ColumnDescriptor;
 import io.vertx.sqlclient.impl.RowDesc;
 
-import java.util.Arrays;
+import java.util.AbstractList;
 import java.util.Collections;
-import java.util.stream.Collectors;
-import java.util.stream.Stream;
+import java.util.List;
+import java.util.RandomAccess;
 
 public class MySQLRowDesc extends RowDesc {
 
-  private final ColumnDefinition[] columnDefinitions;
+  public final ColumnDefinition[] columnDefinitions;
   private final DataFormat dataFormat;
 
-  public MySQLRowDesc(ColumnDefinition[] columnDefinitions, DataFormat dataFormat) {
-    super(Collections.unmodifiableList(Stream.of(columnDefinitions)
-      .map(ColumnDefinition::name)
-      .collect(Collectors.toList())), Collections.unmodifiableList(Arrays.asList(columnDefinitions)));
+  private MySQLRowDesc(List<String> columnNames, List<ColumnDescriptor> columnDescriptors, ColumnDefinition[] columnDefinitions, DataFormat dataFormat) {
+    super(columnNames, columnDescriptors);
     this.columnDefinitions = columnDefinitions;
     this.dataFormat = dataFormat;
   }
 
-  public ColumnDefinition[] columnDefinitions() {
-    return columnDefinitions;
+
+  public static MySQLRowDesc create(ColumnDefinition[] columnDefinitions, DataFormat dataFormat) {
+    if (columnDefinitions.length == 0) {
+      return new MySQLRowDesc(Collections.emptyList(), Collections.emptyList(), columnDefinitions, dataFormat);
+    }
+    List<String> columnNames = new ColumnNames(columnDefinitions);
+    List<ColumnDescriptor> columnDescriptors = new ColumnDescriptors(columnDefinitions);
+    return new MySQLRowDesc(columnNames, columnDescriptors, columnDefinitions, dataFormat);
+  }
+
+  public int size() {
+    return columnDefinitions.length;
+  }
+
+  public ColumnDefinition get(int index) {
+    return columnDefinitions[index];
   }
 
   public DataFormat dataFormat() {
     return dataFormat;
   }
+
+  private static class ColumnNames extends AbstractList<String> implements RandomAccess {
+    private final ColumnDefinition[] columnDefinitions;
+
+    public ColumnNames(ColumnDefinition[] columnDefinitions) {
+      this.columnDefinitions = columnDefinitions;
+    }
+
+    @Override
+    public String get(int index) {
+      return columnDefinitions[index].name();
+    }
+
+    @Override
+    public int size() {
+      return columnDefinitions.length;
+    }
+
+    @Override
+    public int indexOf(Object o) {
+      if (o != null) {
+        for (int i = 0; i < columnDefinitions.length; i++) {
+          String name = columnDefinitions[i].name();
+          if (o.hashCode() == name.hashCode() && o.equals(name)) {
+            return i;
+          }
+        }
+      }
+      return -1;
+    }
+  }
+
+  private static class ColumnDescriptors extends AbstractList<ColumnDescriptor> {
```

Comment:
eventually this should be moved to `RowDesc` is a follow up PR

Reply:
In fact, I may do this when applying changes to other clients.